server/src/Icepeak/Server/WebsocketServer/Payload.hs

```diff
@@ -0,0 +1,274 @@
+instance Aeson.ToJSON ResponseUnsubscribeFailure where
+  toJSON (ResponseUnsubscribeFailure code message mbPaths extraData) =
+    Aeson.object $ baseKeyValues <> addonKeyValues
+    where
+      baseKeyValues :: [ Aeson.Pair ]
+      baseKeyValues =
+        [ "type"  .= ("unsubscribe" :: Text)
+        , "code" .= (code :: Int)
+        , "message" .= (message :: Text)
+        , "extra" .= extraData
+        ]
+
+      addonKeyValues :: [ Aeson.Pair ]
+      addonKeyValues =
+        case mbPaths of
+          Just paths -> [ "paths" .= paths ]
+          Nothing  -> []
+
+
+-- * Close Conn Reason
+
+data CloseType
+  = TypeSizeOutOfBounds
+  | TypeBinaryPayload
+  | TypeJsonDecodeError
+  | TypePayloadNotObject
+  | TypeMissingOrUnexpectedType
+  deriving (Show, Eq)
+
+-- https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/close#parameters
+
+-- Status codes in the range 3000-3999 are reserved for use by
+-- libraries, frameworks, and applications. The interpretation of these codes
+-- is undefined by this protocol.
+
+closeCode :: CloseType -> Word16
+closeCode TypeSizeOutOfBounds         = 3001
+closeCode TypeBinaryPayload           = 3002
+closeCode TypeJsonDecodeError         = 3003
+closeCode TypePayloadNotObject        = 3004
+closeCode TypeMissingOrUnexpectedType = 3005
+
+
+-- A string providing a custom WebSocket connection close reason (a concise human-readable prose explanation for the closure).
+-- The value must be no longer than 123 bytes (encoded in UTF-8).
+
+closeMessage :: CloseType -> Text
+closeMessage TypeSizeOutOfBounds
+  = "Received a payload size that is out of bounds"
+closeMessage TypeBinaryPayload
+  = "Received a payload using Binary instead of Text"
+closeMessage TypeJsonDecodeError
+  = "Received a payload that resulted in a JSON decode error"
+closeMessage TypePayloadNotObject
+  = "Received a JSON payload which is not an object"
+closeMessage TypeMissingOrUnexpectedType
+  = "Received a payload with a missing or unexpected value for 'type'"
+
+-- * Parsing
+
+maxPayloadBytes :: Int
+maxPayloadBytes = 1_000_000
+
+checkBounds
+  :: Lazy.ByteString
+  -> Either RequestMalformedError Lazy.ByteString
+checkBounds lazyBS = isOutOfBoundsAcc lazyBS 0
+  where
+    isOutOfBoundsAcc Lazy.ByteString.Empty _ = Right lazyBS
+    isOutOfBoundsAcc (Lazy.ByteString.Chunk chunk rest) accSize =
+      let accSize' = accSize + ByteString.length chunk in
+        if accSize' > maxPayloadBytes
+        then Left PayloadSizeOutOfBounds
+        else isOutOfBoundsAcc rest accSize'
+
+parseDataMessage
+  :: WebSockets.DataMessage -> RequestPayload
+parseDataMessage (WebSockets.Binary _ ) = RequestPayloadMalformed UnexpectedBinaryPayload
+parseDataMessage (WebSockets.Text utf8EncodedLazyByteString _ ) =
+  case parsedPayload of
+    (Left malformed) -> RequestPayloadMalformed malformed
+    (Right (Left subscribe)) -> RequestPayloadSubscribe subscribe
+    (Right (Right unsubscribe)) -> RequestPayloadUnsubscribe unsubscribe
```

Comment:
I'm not so sure using `Either` alleviates the feeling of unnecessary wrapping.

Using Either keeps the amount of wrapping is the same, but it introduces a layer of nesting:
- https://github.com/channable/icepeak/commit/93fc35024c33afba8bd5af0851670b6f6cff3e45

Reply:
What https://github.com/channable/icepeak/commit/93fc35024c33afba8bd5af0851670b6f6cff3e45 misses is to also remove the `ActionError` case in `determinePayloadAction`.  So changing the type signature to something akin to

```haskell
determinePayloadAction :: Config -> Either RequestMalformedError RequestPayload -> Either RequestMalformedError PayloadAction
``` 

We don't touch `Left` so this can then be simplified to the original 

```haskell
determinePayloadAction :: Config -> RequestPayload -> PayloadAction
```

Which neatly ties this up as given a valid payload, we can create a valid action (via `fmap determinePayloadAction` so it doesn't even touch the malformed payloads).

Reply:
Where should the `RequestMalformedError` be handled/turned into an ActionError?
Are we saying `ActionError` should be removed and just directly dispatched to the action from the `RequestMalformedError`?

This:
```
performPayloadAction
  :: Client -> PayloadAction -> IO ()
```
would become this?
```
performPayloadAction
  :: Client -> Either RequestMalformedError PayloadAction -> IO ()
```